### PR TITLE
Fix MODIS cviirs-based interpolation

### LINKED ISSUE #39: MODIS Interpolation Comparisons
While working on my 3 PRs for optimizing MODIS interpolation (#35, #36, #38), I noticed that in real world cases my cython-based PRs were producing invalid/bad images where the bow-tie pixels of an image were clearly being mapped incorrectly from the default interpolation used by Satpy (the `geotiepoints/modisinterpolator.py` version that uses the CVIIRS algorithm). This happens even though the unit tests pass! This means that the tests aren't strict/rigid enough in their thresholds to the point that they let invalid/bad results through.

NOTE: All plots/images in this issue were made with the current main branch and don't use my updated code.

I've since been trying to do a deep dive on what is going on and how the "simple" interpolation compares to the CVIIRS interpolation. I've discovered at least one problem with the last column of the data. Here is a zoomed in view of the differences between the two methods' longitude arrays, specifically the upper right corner, when used on real Aqua data:

image

You may be able to see on the right most columns (where I've removed the black border line in the matplotlib plot) that the differences are suddenly much higher for the last column. If I look at the individual plots for these longitudes I can't see anything obvious in either method's results. I'll post the images I've generated so far to start the discussion, but we need to come up with ways to decide which method to use as "truth".

Note I refer to the CVIIRS algorithm as "angle-based" because it requires solar zenith angles. I refer to the other method as "simple" because it pretty close to basic bilinear interpolation from the basic understanding I have.

geotiepoints_interp_anglebased
geotiepoints_interp_differences
geotiepoints_interp_simple

CC @simonrp84

### PR DESCRIPTION
**Pull request by djhoese**
Closes #39. Includes #40. Fixes a swath width issue in the cviirs-based interpolation. This PR also recreates the modis interpolation test data to use a less terrain-y dataset. This allows for tighter tolerances in the tests. The tests were also rewritten to use geodetic distance between actual and expected results which is an overall better test than pure value differences.

TODO: Going to rewrite the cviirs and simple tests to use the same utilities.

 - [ ] Closes #xxxx 
 - [ ] Tests added 
 - [ ] Tests passed 
 - [ ] Passes ``git diff origin/main **/*py | flake8 --diff`` 
 - [ ] Fully documented

### CODE REVIEW

**Review by mraspaud — COMMENTED** [review]
Looks good so far, the test data is bigger though, is this planned?

File: geotiepoints/tests/test_modisinterpolator.py [hunk 1]
```diff
@@ -36,6 +37,29 @@ def to_da(arr):
...
+        lons_actual: Longitude array produced by interpolation being tested.
+        lats_actual: Latitude array produced by interpolation being tested.
+        lons_desired: Longitude array of expected/truth coordinates.
+        lats_desired: Latitude array of expected/truth coordinates.
+        max_distance_diff: Limit of allowed distance difference in meters.
+
+    """
+    g = Geod(ellps="WGS84")
+    _, _, dist = g.inv(lons_actual, lats_actual, lons_desired, lats_desired)
+    print(dist.min(), dist.max())
```
[the remark this answers is not in the document]
Comment (djhoese) [reply geotiepoints/tests/test_modisinterpolator.py:59]: Nope. Left over for sanity checking. Thanks.

**Comment by djhoese** [comment]
The test file difference is very confusing. Both are compressed and therefore chunked (automatically done when compressing).

Here is the old 250m latitude:

```
        double lon_250m(phony_dim_2, phony_dim_3) ;
                lon_250m:_Storage = "chunked" ;
                lon_250m:_ChunkSizes = 5, 677 ;
                lon_250m:_Filter = "6,0,3,3385,1,8,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0|1,9" ;
                lon_250m:_Endianness = "little" ;
                lon_250m:_NoFill = "true" ;
```

And new:

```
        float lat_250m(phony_dim_2, phony_dim_3) ;
                lat_250m:_Storage = "chunked" ;
                lat_250m:_ChunkSizes = 10, 677 ;
                lat_250m:_DeflateLevel = 9 ;
                lat_250m:_Endianness = "little" ;
                lat_250m:_NoFill = "true" ;
```

I'll have to do some research why the filtering is being added.

**Comment by djhoese** [comment]
There. Added the shuffle filter and now it is less than 10kb bigger.

**Comment by djhoese** [comment]
Well we have at least one case where running on my system the simple interpolation is <26m max distance, but on CI it is larger (1 pixel).

File: geotiepoints/tests/test_modisinterpolator.py [hunk 2]
```diff
@@ -29,86 +34,136 @@
...
+        lons_actual: Longitude array produced by interpolation being tested.
+        lats_actual: Latitude array produced by interpolation being tested.
+        lons_desired: Longitude array of expected/truth coordinates.
+        lats_desired: Latitude array of expected/truth coordinates.
+        max_distance_diff: Limit of allowed distance difference in meters.
+
+    """
+    g = Geod(ellps="WGS84")
+    _, _, dist = g.inv(lons_actual, lats_actual, lons_desired, lats_desired)
+    np.testing.assert_array_less(dist, max_distance_diff)  # meters
```
Comment (mraspaud) [inline geotiepoints/tests/test_modisinterpolator.py:119]: For clarity, I think we should provide a custom error message to `assert_array_less`.

**Review by mraspaud — COMMENTED** [review]
Just a couple of minor comments, otherwise looks good to me.

File: geotiepoints/tests/test_modisinterpolator.py (hunk 2 above)
Comment (djhoese) [reply geotiepoints/tests/test_modisinterpolator.py]:
```
        err_msg=f"Coordinates are greater than {max_distance_diff} geodetic "
                "meters from the expected coordinates.")
```

Done.

File: geotiepoints/tests/test_modisinterpolator.py [hunk 3]
```diff
@@ -29,86 +34,136 @@
...
+    assert not np.any(np.isnan(lats))
+
+
+def test_cviirs_nan_handling():
+    # See GH #19
+    lon1, lat1, satz1 = load_1km_lonlat_satz_as_xarray_dask()
+    satz1 = _to_da(abs(np.linspace(-65.4, 65.4, 1354)).repeat(20).reshape(-1, 20).T)
+    lons, lats = modis_1km_to_500m(lon1, lat1, satz1)
+    assert not np.any(np.isnan(lons.compute()))
+    assert not np.any(np.isnan(lats.compute()))
```
[the remark this answers is not in the document]
Comment (djhoese) [reply geotiepoints/tests/test_modisinterpolator.py]: Changed it in these test names. Is that what you wanted?

**Comment by djhoese** [comment]
I think since I have commits here with multiple HDF5 test data versions we should squash this PR.

**Pull request merged**